pkg/proxy/iptables/proxier_test.go

```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+			return false
+		}
+	}
+
+	// Ignore SessionAffinity rules
+	// affinityRegex = / --rcheck /
+	if affinityRegex.MatchString(rule) {
+		return false
+	}
+
+	// Anything else is assumed to match
+	return true
+}
+
+// runChain runs the given packet through the rules in the given table and chain, updating
+// tracer's internal state accordingly. It returns true if it hits a terminal action.
+func (tracer *iptablesTracer) runChain(table, chain, sourceIP, destIP, destPort string) bool {
+	for _, rule := range tracer.tables[table][chain] {
+		// Match any rule that jumps to another chain or built-in target.
+		// jumpRegex = /(--probability.*)? -j ([^ ]*)( --to-destination (.*))?$/
+		match := jumpRegex.FindStringSubmatch(rule)
+		if match == nil {
+			continue
+		}
+		isProbabilisticMatch := (match[1] != "")
+		target := match[2]
+		natDestination := match[4]
+
+		if !tracer.ruleMatches(rule, sourceIP, destIP, destPort) {
+			continue
+		}
+		tracer.matches = append(tracer.matches, rule)
+
+		switch target {
+		case "KUBE-MARK-MASQ":
+			tracer.markMasq = true
+			continue
+
+		case "KUBE-MARK-DROP":
+			tracer.markDrop = true
+			continue
+
+		case "ACCEPT", "REJECT":
+			tracer.outputs = append(tracer.outputs, target)
+			return true
+
+		case "DNAT":
+			tracer.outputs = append(tracer.outputs, natDestination)
+			return true
+
+		default:
+			// We got a "-j KUBE-SOMETHING", so process that chain
+			terminated := tracer.runChain(table, target, sourceIP, destIP, destPort)
+
+			// If the subchain hit a terminal rule AND the rule that sent us
+			// to that chain was non-probabilistic, then this chain terminates
+			// as well. But if we went there because of a --probability rule,
+			// then we want to keep accumulating further matches against this
+			// chain.
+			if terminated && !isProbabilisticMatch {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// tracePacket determines what would happen to a packet with the given sourceIP, destIP,
+// and destPort, given the indicated iptables ruleData. nodeIP is the local node IP (for
+// rules matching "LOCAL").
+//
+// The return values are: an array of matched rules (for debugging), the final packet
+// destinations (a comma-separated list of IPs, or one of the special targets "ACCEPT",
+// "DROP", or "REJECT"), and whether the packet would be masqueraded.
+func tracePacket(ruleData, sourceIP, destIP, destPort, nodeIP string) ([]string, string, bool) {
+	tracer, err := newIPTablesTracer(ruleData, nodeIP)
+	if err != nil {
+		panic("bad iptables ruleData: " + err.Error())
+	}
+
+	// nat:PREROUTING goes first, then the filter chains, then nat:POSTROUTING. For
+	// our purposes that means we run through the "nat" chains first, starting at
+	// KUBE-SERVICES, then we do the "filter" chains. The only interesting thing that
+	// happens in nat:POSTROUTING is that the masquerade mark gets turned into actual
+	// masquerading.
+
+	tracer.runChain("nat", "KUBE-SERVICES", sourceIP, destIP, destPort)
```

Comment:
comment that the jump from PREROUTING to KUBE-SERVICES is configured outside of this logic, so that's why we're manually doing it here